c2/cuda/groth16_cuda.cu

```diff
@@ -111,6 +120,10 @@ RustError::by_value generate_groth16_proofs_c(const Assignment<fr_t> provers[],
     static std::mutex mtx;
     std::lock_guard<std::mutex> lock(mtx);
 
+    if (!ngpus()) {
+        return RustError{-9, "No CUDA devices available"};
```

Comment:
Suggestion also is to reuse error codes as opposed to making up own. For example in this case one can use CUDA_ERROR_NO_DEVICE or ENODEV...